### DISCUSSION TITLE
Perform dataflow analysis in the bytecode compiler

### DISCUSSION THREAD

**Discussion by markshannon** [faster-cpython/ideas#306] [Ideas]
There a number of standard compiler analyses that we could implement and would be useful. For example:
* Liveness analysis: determine when a variable is first defined, redefined and killed.
* Constant propogation: Determine when a variable holds a constant.

Since the 1990s, the conventional way to perform these, and many other, analyses has been [SSA form](https://en.wikipedia.org/wiki/Static_single_assignment_form).
SSA form is a bit intimidating to start with, but once you are used to it, it seems natural and easy to work with.

Dataflow analysis would allow us to do less work handling variables and the stack, avoiding some NULL checks.
For example:
* The first store to a local variable does not need to handle the old value, as it will be NULL.
* Other stores to local variables do not need a NULL check of the old value.
* `LOAD_FAST` will not need a NULL check as the compiler can detect if the variable can be undefined (and insert a check in the rare case that it is necessary)

By themselves, these changes might not justify the extra complexity in the compiler, but combined with the refcount improvements in https://github.com/faster-cpython/ideas/discussions/308, it might.

We would need to take care when modifying local variables in the debugger.
We might have to accept some rare memory leaks, and disallow deletion of local variables (assigning `None` instead).

**Comment by markshannon** [comment]
This is basically impossible with the compiler in its current form.
When https://github.com/faster-cpython/ideas/discussions/235 is complete, it should be feasible.

**Comment by pxeger** [comment]
So the bytecode would be converted into SSA form, optimised, and then converted back into stack-machine bytecode? Or what exactly are you proposing?

**Comment by markshannon** [comment]
The IR is stack based, but it isn't bytecode.

**Reply by penguin-wwy** [reply]
AST -> stack-based SSA IR -> bytecode，Is that what you mean?

**Comment by markshannon** [comment]
Something like that.
The operand of both `LOAD_FAST` and `STORE_FAST` instructions is an index into the local array. Converting to SSA would involve adding an other index. E.g. `LOAD_FAST a` is represented as `LOAD_FAST 1`. Converting to SSA form, `LOAD_FAST a0` would be represented as `LOAD_FAST 1,0`
The resulting CFG would still be stack-based, but in SSA form.

**Comment by sweeneyde** [comment]
I think we can get a first level of liveness analysis with a hundred or so lines of code, without transforming everything into SSA, by just walking through the graph of basic blocks and marking things as "unsafe" when it's possible that they haven't been initialized yet.

I have a branch [here](https://github.com/python/cpython/compare/main...sweeneyde:known?expand=1) that has all tests passing, except `test_dis`. I only added `LOAD_FAST_KNOWN`, since that is the most common opcode.

I have yet to deal with the debugger situation, but I was thinking it could convert every `LOAD_FAST_KNOWN` into the safer `LOAD_FAST` whenever any local variable setter is used or whenever `frame.f_lineno` is set, and just have this be irreversible.

**Reply by sweeneyde** [reply]
From a complete run through the test suite gathering stats, there were 1000x as many LOAD_FAST_KNOWN as LOAD_FAST.

**Comment by sweeneyde** [comment]
PyPerformance somehow got "Geometric mean: 1.01x slower" for me

<details>

```
Slower (17):
- nqueens: 80.5 ms +- 2.2 ms -> 93.6 ms +- 23.3 ms: 1.16x slower
- logging_simple: 6.63 us +- 0.12 us -> 7.41 us +- 1.83 us: 1.12x slower
- mako: 9.83 ms +- 0.10 ms -> 10.9 ms +- 2.0 ms: 1.10x slower
- logging_format: 7.29 us +- 0.16 us -> 8.04 us +- 1.81 us: 1.10x slower
- nbody: 87.7 ms +- 2.2 ms -> 95.5 ms +- 15.9 ms: 1.09x slower
- meteor_contest: 106 ms +- 2 ms -> 116 ms +- 23 ms: 1.09x slower
- unpack_sequence: 40.7 ns +- 0.7 ns -> 43.4 ns +- 0.7 ns: 1.07x slower
- pickle_dict: 25.1 us +- 0.4 us -> 26.3 us +- 0.7 us: 1.05x slower
- pickle_list: 4.13 us +- 0.09 us -> 4.31 us +- 0.09 us: 1.04x slower
- genshi_xml: 58.9 ms +- 3.1 ms -> 60.7 ms +- 2.3 ms: 1.03x slower
- pathlib: 19.7 ms +- 0.4 ms -> 20.3 ms +- 1.2 ms: 1.03x slower
- genshi_text: 22.4 ms +- 0.6 ms -> 23.0 ms +- 0.7 ms: 1.03x slower
- xml_etree_process: 52.0 ms +- 0.6 ms -> 52.9 ms +- 0.7 ms: 1.02x slower
- chameleon: 6.34 ms +- 0.10 ms -> 6.42 ms +- 0.10 ms: 1.01x slower
- regex_dna: 193 ms +- 4 ms -> 195 ms +- 2 ms: 1.01x slower
- python_startup_no_site: 6.61 ms +- 0.05 ms -> 6.65 ms +- 0.05 ms: 1.01x slower
- 2to3: 258 ms +- 3 ms -> 258 ms +- 2 ms: 1.00x slower

Faster (16):
- richards: 51.9 ms +- 2.3 ms -> 48.7 ms +- 2.1 ms: 1.07x faster
- spectral_norm: 93.2 ms +- 1.1 ms -> 88.7 ms +- 1.6 ms: 1.05x faster
- pidigits: 192 ms +- 1 ms -> 185 ms +- 5 ms: 1.03x faster
- scimark_monte_carlo: 65.9 ms +- 1.2 ms -> 63.9 ms +- 0.9 ms: 1.03x faster
- json_loads: 27.9 us +- 1.7 us -> 27.1 us +- 2.0 us: 1.03x faster
- raytrace: 302 ms +- 7 ms -> 293 ms +- 9 ms: 1.03x faster
- regex_v8: 21.7 ms +- 0.3 ms -> 21.2 ms +- 0.2 ms: 1.02x faster
- scimark_sor: 106 ms +- 2 ms -> 104 ms +- 2 ms: 1.02x faster
- telco: 6.45 ms +- 0.15 ms -> 6.34 ms +- 0.18 ms: 1.02x faster
- unpickle_list: 4.87 us +- 0.06 us -> 4.80 us +- 0.13 us: 1.02x faster
- hexiom: 6.36 ms +- 0.11 ms -> 6.27 ms +- 0.11 ms: 1.01x faster
- go: 137 ms +- 2 ms -> 135 ms +- 2 ms: 1.01x faster
- regex_effbot: 2.84 ms +- 0.03 ms -> 2.80 ms +- 0.05 ms: 1.01x faster
- crypto_pyaes: 76.3 ms +- 1.4 ms -> 75.4 ms +- 1.6 ms: 1.01x faster
- dulwich_log: 77.3 ms +- 1.1 ms -> 76.4 ms +- 1.1 ms: 1.01x faster
- fannkuch: 372 ms +- 4 ms -> 370 ms +- 6 ms: 1.01x faster

Benchmark hidden because not significant (26): chaos, deltablue, django_template, float, html5lib, json_dumps, logging_silent, pickle, pickle_pure_python, pyflate, python_startup, regex_compile, scimark_fft, scimark_lu, scimark_sparse_mat_mult, sqlite_synth, sympy_expand, sympy_integrate, sympy_sum, sympy_str, tornado_http, unpickle, unpickle_pure_python, xml_etree_parse, xml_etree_iterparse, xml_etree_generate

Geometric mean: 1.01x slower
```
</details>

**Reply by sweeneyde** [reply]
Re-running, I got "Geometric mean: 1.01x faster". So it looks like it's within the noise of what my machine can reliably detect on pyperformance. But it might be worth it regardless, I'm not sure.

<details>

```
Slower (10):
- pickle: 10.8 us +- 1.1 us -> 11.4 us +- 1.0 us: 1.06x slower
- sympy_integrate: 21.5 ms +- 0.3 ms -> 22.2 ms +- 1.3 ms: 1.03x slower
- pickle_dict: 24.9 us +- 0.6 us -> 25.6 us +- 0.7 us: 1.03x slower
- genshi_xml: 60.2 ms +- 2.4 ms -> 61.4 ms +- 2.9 ms: 1.02x slower
- regex_dna: 193 ms +- 3 ms -> 197 ms +- 3 ms: 1.02x slower
- pathlib: 19.4 ms +- 0.3 ms -> 19.7 ms +- 0.4 ms: 1.01x slower
- pickle_list: 4.16 us +- 0.10 us -> 4.21 us +- 0.14 us: 1.01x slower
- sympy_sum: 175 ms +- 2 ms -> 177 ms +- 5 ms: 1.01x slower
- telco: 6.48 ms +- 0.15 ms -> 6.54 ms +- 0.17 ms: 1.01x slower
- sqlite_synth: 2.37 us +- 0.03 us -> 2.39 us +- 0.04 us: 1.01x slower

Faster (32):
- scimark_fft: 349 ms +- 8 ms -> 326 ms +- 12 ms: 1.07x faster
- nbody: 92.0 ms +- 2.8 ms -> 86.3 ms +- 1.7 ms: 1.07x faster
- logging_silent: 98.8 ns +- 11.0 ns -> 93.9 ns +- 2.4 ns: 1.05x faster
- hexiom: 6.57 ms +- 0.13 ms -> 6.25 ms +- 0.10 ms: 1.05x faster
- scimark_sparse_mat_mult: 4.99 ms +- 0.19 ms -> 4.78 ms +- 0.26 ms: 1.05x faster
- richards: 53.7 ms +- 3.0 ms -> 51.4 ms +- 2.5 ms: 1.05x faster
- django_template: 36.9 ms +- 1.4 ms -> 35.5 ms +- 2.5 ms: 1.04x faster
- json_loads: 27.7 us +- 0.9 us -> 26.7 us +- 1.2 us: 1.04x faster
- spectral_norm: 95.1 ms +- 3.6 ms -> 91.6 ms +- 2.9 ms: 1.04x faster
- pidigits: 193 ms +- 1 ms -> 186 ms +- 2 ms: 1.03x faster
- go: 139 ms +- 4 ms -> 134 ms +- 3 ms: 1.03x faster
- logging_format: 7.58 us +- 0.25 us -> 7.35 us +- 0.15 us: 1.03x faster
- nqueens: 86.2 ms +- 2.0 ms -> 83.6 ms +- 2.7 ms: 1.03x faster
- raytrace: 303 ms +- 8 ms -> 295 ms +- 6 ms: 1.03x faster
- mako: 10.1 ms +- 0.1 ms -> 9.81 ms +- 0.17 ms: 1.03x faster
- fannkuch: 376 ms +- 6 ms -> 366 ms +- 10 ms: 1.03x faster
- chaos: 72.4 ms +- 1.9 ms -> 70.5 ms +- 2.0 ms: 1.03x faster
- xml_etree_process: 53.8 ms +- 1.4 ms -> 52.5 ms +- 1.3 ms: 1.02x faster
- meteor_contest: 109 ms +- 2 ms -> 106 ms +- 2 ms: 1.02x faster
- scimark_monte_carlo: 65.6 ms +- 1.3 ms -> 64.1 ms +- 1.6 ms: 1.02x faster
- logging_simple: 6.79 us +- 0.20 us -> 6.64 us +- 0.15 us: 1.02x faster
- dulwich_log: 77.9 ms +- 1.3 ms -> 76.3 ms +- 1.9 ms: 1.02x faster
- unpack_sequence: 42.3 ns +- 1.1 ns -> 41.5 ns +- 0.9 ns: 1.02x faster
- regex_compile: 142 ms +- 3 ms -> 140 ms +- 3 ms: 1.02x faster
- unpickle_pure_python: 238 us +- 8 us -> 234 us +- 8 us: 1.02x faster
- genshi_text: 22.8 ms +- 0.6 ms -> 22.4 ms +- 0.7 ms: 1.02x faster
- 2to3: 261 ms +- 9 ms -> 257 ms +- 3 ms: 1.02x faster
- xml_etree_generate: 76.0 ms +- 2.2 ms -> 75.1 ms +- 1.5 ms: 1.01x faster
- regex_effbot: 2.87 ms +- 0.09 ms -> 2.83 ms +- 0.06 ms: 1.01x faster
- scimark_lu: 101 ms +- 3 ms -> 99.9 ms +- 2.2 ms: 1.01x faster
- crypto_pyaes: 76.7 ms +- 1.8 ms -> 76.0 ms +- 1.3 ms: 1.01x faster
- python_startup: 8.95 ms +- 0.06 ms -> 8.87 ms +- 0.05 ms: 1.01x faster

Benchmark hidden because not significant (17): chameleon, deltablue, float, html5lib, json_dumps, pickle_pure_python, pyflate, python_startup_no_site, regex_v8, scimark_sor, sympy_expand, sympy_str, tornado_http, unpickle, unpickle_list, xml_etree_parse, xml_etree_iterparse

Geometric mean: 1.01x faster
```
</details>

**Comment by markshannon** [comment]
NULL checks are cheap: it is a perfectly predicted branch, testing a value that it is in a register. So I'm not surprised the difference in performance is negligible. 

Also, I don't think removing the NULL check is safe. Potentially, someone could delete a local in the debugger.
Maybe we should disallow such deletions, but that needs something like PEP 667.

**Reply by sweeneyde** [reply]
It wouldn't be safe with no other alterations, but I also changed `frame_setlineno` (always) and `_PyFrame_LocalsToFast` (on deletions) to convert all the unsafe-ness back to plain old `LOAD_FAST`. Are there more sites where unsafe-ness could arise?

**Comment by sweeneyde** [comment]
I ran pyperformance 3 times on main, 3 times on the branch with the `LOAD_FAST_KNOWN`, and 3 times on [a branch](https://github.com/python/cpython/compare/main...sweeneyde:microopt) that also added a couple more micro-optimizations to `LOAD_FAST`: not setting `frame->prev_instr` and also using `NOTRACE_DISPATCH`. I measured a consistent 1% speedup from main to the branch with those 3 micro-optimizations.

https://gist.github.com/sweeneyde/32b69bea8f7e15df7bbd9c41eac85f8d

**Reply by sweeneyde** [reply]
Some microbenchmarks seem to budge a few percent as well:

Script:
<details>

```python
from pyperf import Runner, perf_counter
from itertools import repeat

runner = Runner()

def bench(func):
    runner.bench_time_func(func.__name__, func, inner_loops=100)
    return func

@bench
def LOAD_FAST(loops):
    rep = repeat(None, loops)
    x = None
    t0 = perf_counter()
    for _ in rep:
        x;x;x;x;x; x;x;x;x;x; x;x;x;x;x; x;x;x;x;x;
        x;x;x;x;x; x;x;x;x;x; x;x;x;x;x; x;x;x;x;x;
        x;x;x;x;x; x;x;x;x;x; x;x;x;x;x; x;x;x;x;x;
        x;x;x;x;x; x;x;x;x;x; x;x;x;x;x; x;x;x;x;x;
        x;x;x;x;x; x;x;x;x;x; x;x;x;x;x; x;x;x;x;x;

    return perf_counter() - t0


@bench
def STORE_FAST__LOAD_FAST(loops):
    rep = repeat(None, loops)
    x = None
    t0 = perf_counter()
    for _ in rep:
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
        x=x;x=x;x=x;x=x;x=x; x=x;x=x;x=x;x=x;x=x;
 
    return perf_counter() - t0

@bench
def LOAD_FAST__LOAD_FAST(loops):
    rep = repeat(None, loops)
    x = None
    t0 = perf_counter()

    for _ in rep:
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;
        x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x; x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;x,x=x,x;

    return perf_counter() - t0

@bench
def LOAD_FAST__LOAD_CONST(loops):
    rep = repeat(None, loops)
    x = None
    t0 = perf_counter()

    for _ in rep:
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;
        x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1; x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;x,x=x,1;

    return perf_counter() - t0

@bench
def LOAD_CONST__LOAD_FAST(loops):
    rep = repeat(None, loops)
    x = None
    t0 = perf_counter()

    for _ in rep:
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;

        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x; x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;x,x=1,x;
        
    return perf_counter() - t0
```
</details>

Result:

| Benchmark             | before_micro  | LOAD_FAST_KNOWN       | LOAD_FAST_KNOWN+more_microopt  |
|-----------------------|:-------------:|:---------------------:|:------------------------------:|
| LOAD_FAST             | 2.94 ns       | 2.80 ns: 1.05x faster | 2.74 ns: 1.07x faster          |
| STORE_FAST__LOAD_FAST | 3.20 ns       | 3.10 ns: 1.03x faster | 3.05 ns: 1.05x faster          |
| LOAD_FAST__LOAD_FAST  | 6.04 ns       | 5.60 ns: 1.08x faster | 5.55 ns: 1.09x faster          |
| LOAD_FAST__LOAD_CONST | 4.49 ns       | 4.06 ns: 1.11x faster | 4.13 ns: 1.09x faster          |
| LOAD_CONST__LOAD_FAST | 5.46 ns       | 5.48 ns: 1.00x slower | 5.54 ns: 1.01x slower          |
| Geometric mean        | (ref)         | 1.05x faster          | 1.06x faster                   |

**Comment by markshannon** [comment]
Rather than add a `LOAD_FAST_KNOWN` I would suggest add a `CHECK_LOCAL` opcode which does the `NULL` check and remove the `NULL` check from `LOAD_FAST`. That way, we can keep the current superinstructions.

**Reply by sweeneyde** [reply]
At least without higher-tier optimizers, wouldn't that add a lot of extra dispatch overhead to our most common opcode? Besides, we can modify the current superinstructions so that `LOAD_FAST__LOAD_FAST` is a actually `LOAD_FAST_KNOWN__LOAD_FAST_KNOWN` without much loss in how broadly applicable those superinstructions are. As I mentioned, 99.9% of all `LOAD_FAST` can become `LOAD_FAST_KNOWN`.

My draft [here](https://github.com/python/cpython/pull/92763) that implements this.

**Reply by markshannon** [reply]
You say that 99.9% of `LOAD_FAST`s don't need a `NULL` check, so why would adding `CHECK_LOCAL` opcode for the remaining 0.1% add extra dispatch overhead?

**Reply by sweeneyde** [reply]
Ah, I misunderstood -- that would make sense. But then I wouldn't know how to re-insert those checks for cases where locals are deleted or f_lineno is set.

**Reply by markshannon** [reply]
Ah yes, so we do need to a checked form of `LOAD_FAST`. Your approach looks good then.

To avoid having to rename all the superinstructions, and so that the common form has the shorter name, I'd rename `LOAD_FAST_KNOWN` as `LOAD_FAST` and rename the checked version as `LOAD_FAST_CHECK` or similar.

**Comment by markshannon** [comment]
With regard to your draft implementation. 99.9% fewer NULL checks is excellent.
I think this would be better done in the compiler. The compiler can emit a `CHECK_FAST` before the `LOAD_FAST` when we can't prove that the local won't be NULL. This allows existing code for superinstructions to work unchanged.

Regarding removal of the `frame->prev_instr = next_instr;` assignment. That is probably best left until we generate the interpreter, as we can apply it to:
* LOAD_FAST
* LOAD_CONST
* Super instructions of the above
* RESUME_QUICK
* RETURN_VALUE
and maybe a couple of others, but we don't want to do it manually.

**Comment by sweeneyde** [comment]
https://github.com/python/cpython/pull/93144

**Comment by markshannon** [comment]
We could do something like https://github.com/python/cpython/pull/93144 for stores.

From analysis, we should know (for most cases at least) whether the variable was `NULL` or not before the store.
To take advantage of this info we would want three forms of `STORE_FAST`. Something like:
* `STORE_FAST_CLEARED`. No need to read the old value, just store the new one.
* `STORE_FAST_REPLACE`. Must decref the old value, but don't need to check if it was `NULL`
* `STORE_FAST`. The current implementation, needed to handle `del frame.f_locals["x"]`, etc. 

This would convert one instruction into three, and would negatively impact the superinstructions, so it isn't clear that it would be worthwhile.
But definitely worth investigating, IMO.

**Reply by sweeneyde** [reply]
[Here are stats results](https://gist.github.com/sweeneyde/b062aee76d65dcb7c7e989749c541800) for `python -m test` after including something like those three and disabling all super-instructions. [Branch is here.](https://github.com/python/cpython/compare/main...sweeneyde:known_stores?expand=1) In particular:

|Name | Count | Self | Cumulative | Miss ratio | 
|---|---:|---:|---:|---:|
| STORE_FAST_GENERAL | 237282371 | 4.1% | 36.7% |  |
| STORE_FAST_INIT | 152727216 | 2.7% | 48.8% |  |
| STORE_FAST_REPLACE | 66206667 | 1.2% | 71.9% |  |

Also:

<details>
<summary> Successors and predecessors for STORE_FAST_GENERAL </summary>

|Predecessors | Count | Percentage | 
|---|---:|---:|
| FOR_ITER | 85321161 | 36.0% |
| STORE_FAST_GENERAL | 52104997 | 22.0% |
| UNPACK_SEQUENCE_TWO_TUPLE | 32921056 | 13.9% |
| RETURN_VALUE | 12808786 | 5.4% |
| CALL_ADAPTIVE | 11076751 | 4.7% |

|Successors | Count | Percentage | 
|---|---:|---:|
| LOAD_FAST | 110005042 | 46.4% |
| STORE_FAST_GENERAL | 52104997 | 22.0% |
| LOAD_CONST | 16911415 | 7.1% |
| LOAD_DEREF | 11810924 | 5.0% |
| LOAD_GLOBAL_MODULE | 11185232 | 4.7% |

</details>

I'd bet loop-peeling the first iteration of every for-loop would shrink the number of STORE_FAST_GENERAL. But I'm not sure that's worth it for now.

**Comment by markshannon** [comment]
Thanks for trying that out.

I think we'll leave loop peeling for the higher tier optimizers (it happens implicitly if the VM is looking for hot backedges).
